### ISSUE TITLE
Regarding HCRT_BOOTSTRAP.BIN

### ISSUE THREAD

**Issue by y4my4my4m** [nrootconauto/EXODUS#7]
What is the content of HCRT_BOOTSTRAP.BIN?
Can I build it from source?

I'm trying to make a ZealOS port, but I'm finding difficult because it seems to try to refer files that have moved/renamed.

**Comment by y4my4my4m** [comment]
I found this in the CONTRIBUTING.md :eyes:
`Build HCRT.BIN using ./exodus -c -t T BuildHCRT.HC, copy HCRT.BIN to HCRT_BOOTSTRAP.BIN`

**Comment by y4my4my4m** [comment]
Do I need HCRT.BIN to run BuildHCRT ?

**Comment by y4my4my4m** [comment]
What is vec/vec.h and map/map.h in src/exodus/loader.h?

**Comment by y4my4my4m** [comment]
So, since ZealOS is somewhat heavily modified, many of the calls have differed. I've tried to adapt one to the other (renaming everything accordingly), however it seems that I need HCRT to build HCRT but because the Kernel has changed definitions (for example GET_RAX is now RAX_GET), it seems to be incompatible with the previously built HCRT.BIN

How do I go about creating one from scratch without using an HCRT file? Do I load Zeal in a VM and try to run it and use the resulting file?

**Comment by y4my4my4m** [comment]
Welp, that's about as far as I could do today 
https://github.com/y4my4my4m/EZODUS

**Comment by nrootconauto** [comment]
@y4my4my4m The most nefarious change I did with HCRT.BIN is using a "Host" IC_FS. I basically rooted out the ring0 stuff too. At one point you could make a TOS  distro from EXODUS but my freind changed the IC_FS(which was previously a polyfill). I will spend tonight trying to port ZealOS.

EXODUS has a complicated histroy,it started from my old 3Days repo,which I basically discarded after the bootstrapping compiler could build the HCRT.BIN . Its chicken and egg. I will try to port ZealOS to EXODUS. Just give me a few days

**Comment by y4my4my4m** [comment]
@nrootconauto thank you so much! I really appreciate it. Feel free to ask questions and I'll answer them as best I can.

Things haven't changed that much so in theory it should be straightforward.

Some changes include:
- functions renaming (verb comes last)
- files renamed/splitting
- SSE instructions added
- supports multiple resolution (still defaults to 640x480 to obey the charter)

You can check my fork EZODUS's for the HCTR_TOS.ZC to get a sense of where those files now are.

**Comment by degtine** [comment]
Im the freind he mentioned, you'll want to check out [Compiler/OPTPASS789A.HC#L422](https://github.com/nrootconauto/EXODUS/blob/869f993521ca5aaba8b32f5b6224a687fda3a214/T/Compiler/OPTPASS789A.HC#L422).You can see how the FS:0x..[RAX] we use to access Fs(and GS for Gs) in assembly isnt posible due to this,this change is because the operating systems are already either using FS or GS. We use [GetFs](https://github.com/nrootconauto/EXODUS/blob/869f993521ca5aaba8b32f5b6224a687fda3a214/T/HCRT_TOS.HC#L18) ([used here](https://github.com/nrootconauto/EXODUS/blob/869f993521ca5aaba8b32f5b6224a687fda3a214/T/Kernel/Sched.HC#L59)) instead which isn't perfect but does the job.

It used to just be an FFI that calls C thread-local storeage,but this was really slow on Windows because it uses "emulated" TLS instead of "real" TLS. This is the major difference between TempleOS and Exodus which makes HCRT compiled from TempleOS not work.You can change it back to a normal FFI function, but you'll need to put paratheses around & like `&(Fs->except_ch)` because normal functions have slightly different operator precedence. I think your best bet would be to change stuff out until it's ZealOS.

**Comment by y4my4my4m** [comment]
Seems like I really can't get this to work.
@nrootconauto did you ever get a chance at looking into it?

**Comment by nrootconauto** [comment]
Yes,I am looking at it. I will write a tool to look for Exodus imports then I will grep for all of them throughtout Zeal Code
The most important things to look at will be 
- MAlloc and freinds
- BlkDev(For FSt_VIRT)
- Graphics.
I finaly got around to doing things in the spare hours of night,Ill keep you posted here

**Comment by y4my4my4m** [comment]
@nrootconauto thank you, i really appreciate it :pray:

**Comment by nrootconauto** [comment]
I have been working on the EZODUS. I am using a `#ifdef EZODUS` so the changes in the source code. For the `_intern` functions did you change the ordering any of them(I am bit sick now otherwise I would look myself).

My basic plan is to Generate a gaint headers file,as there will be 1 ZCRT.BIN file(the way I do it in exodus). I think this is ok. In TempleOS is basically compiles the compiler and kernel separately which would be weird for what im working on(because I load in the symbols from C after `Load`ing them).

Ill be able to work more on this in the coming days. All Ezodus specific code will have its own directory to not deviate too much from normal ZealOS code

**Comment by y4my4my4m** [comment]
@nrootconauto 
the `_intern` where? In the kernel? 
A few things have changed but the order is mostly the same, however sometimes the files have moved location or have been split in parts. Some additions have been made though for example SSE instruction support.

Some useful files to look at: 
[ConversionScript](https://github.com/Zeal-Operating-System/ZealOS/blob/b08af238d0aab92a80b018d8b5fcf80c1e2d0800/src/System/Utils/ConversionScript.ZC) (not quite but almost complete change list of variable/function names)

[HRTC_TOS.ZC](https://github.com/y4my4my4m/EZODUS/blob/173d4d021b8f29374ddbfe2cf9ed4103121ce060/T/HCRT_TOS.ZC#L70-L152) i believe this correctly renamed the include list (hopefully i didnt miss or misunderstand anything)

**Comment by y4my4my4m** [comment]
@nrootconauto let me know if there's anything that is confusing about the nature of the changes on zealos btw 🙏

**Comment by y4my4my4m** [comment]
@nrootconauto any update?
